core/src/main/java/org/kohsuke/stapler/WebApp.java

```diff
@@ -224,13 +224,19 @@ public MetaClass getMetaClass(Class c) {
     
     public MetaClass getMetaClass(Klass<?> c) {
         if(c==null)     return null;
-        synchronized(classMap) {
-            MetaClass mc = classMap.get(c);
-            if(mc==null) {
-                mc = new MetaClass(this,c);
-                classMap.put(c,mc);
+        if (c.navigator == KlassNavigator.JAVA) {
+            if (classMap == null) {
+                classMap = new ClassValue<MetaClass>() {
+                    @Override
+                    protected MetaClass computeValue(Class<?> c) {
+                        return new MetaClass(WebApp.this, Klass.java(c));
+                    }
+                };
```

Comment:
Presumably we are not concerned about concurrent callers because this is just a cache? Could use [double-checked locking](https://errorprone.info/bugpattern/DoubleCheckedLocking) here to eliminate the possibility of concurrent callers getting different cached values, which I think is a nice property to have in theory but likely does not matter much in practice.

Reply:
`ClassValue` claims to have its own concurrency control. I do not think it matters here anyway since while these objects are not immutable, they do not hold valuable state.

Reply:
Yeah, I know that `ClassValue` is thread-safe; my point was that two concurrent callers to `getMetaClass` could wind up creating two separate instances of `ClassValue` (the second one clobbering the first) and thus the first instance of `ClassValue` could compute one value for a given key while the second instance of `ClassValue` could compute another value for the same key. As I wrote originally, probably does not matter in practice, but double-checked locking would be a clean way to make this "tighter" in theory. No strong preference either way.

Reply:
Oh I see, you are concerned about https://github.com/jenkinsci/stapler/blob/0c572f5f289004ad7fd725e8c89cbba827a2f70c/core/src/main/java/org/kohsuke/stapler/WebApp.java#L123, not `ClassValue` behavior per se. I think harmless but it should be easy enough to prevent that.

Reply:
Indeed. The version from https://errorprone.info/bugpattern/DoubleCheckedLocking avoids `synchronized` in the common case (getting the volatile field after it has already been initialized) but this is definitely good enough, no need to over-optimize.